File: 백준/Gold/1245. 농장 관리/농장 관리.py
# 산봉우리
import sys
input = sys.stdin.readline

def dfs(x, y, h):

    ret = True
    visited[x][y] = 1

    for dx, dy in direc:
        nx, ny = x + dx, y + dy

        if nx < 0 or ny < 0 or nx >= N or ny >= M:
            continue
    
        if graph[nx][ny] == h and not visited[nx][ny]:
            ret &= dfs(nx, ny, h)
            
        elif graph[nx][ny] > h:
            ret = False

    return ret

N, M = map(int,input().split())
visited = [[0] * M for _ in range(N)]
cnt = 0
direc = [(1, 0), (-1, 0), (0, 1), (0, -1),
         (1, 1), (1, -1), (-1, 1), (-1, -1)]

graph = []
for i in range(N):
    graph.append(list(map(int,input().split())))

for i in range(N):
    for j in range(M):
        
        if visited[i][j]:
            continue

        if dfs(i, j, graph[i][j]):
            cnt += 1

print(cnt)

File: 백준/Gold/1245. 농장 관리/농장 관리.java
// 산봉우리
import java.io.*;
import java.util.*;

public class Main {

    static int N, M;
    static int[][] graph;
    static int[][] visited;
    static int[][] direc = {
        {1, 0}, {-1, 0}, {0, 1}, {0, -1},
        {1, 1}, {1, -1}, {-1, 1}, {-1, -1}
    };

    // 높이가 같은 지역은 탐색
    static boolean dfs(int x, int y, int h) {

        boolean ret = true;
        visited[x][y] = 1;

        for (int[] d : direc) {
            int nx = x + d[0];
            int ny = y + d[1];

            if (nx < 0 || ny < 0 || nx >= N || ny >= M)
                continue;

            // 높이가 같은 지역은 탐색
            if (graph[nx][ny] == h && visited[nx][ny] == 0)
                ret &= dfs(nx, ny, h);
            // 주변에 더 높은 지역이 있다면 False로 만들어준다
            else if (graph[nx][ny] > h)
                ret = false;
        }

        return ret;
    }

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        visited = new int[N][M];
        graph = new int[N][M];
        int cnt = 0;
        // 인점한 방향 이동 거리

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                graph[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {

                if (visited[i][j] == 1)
                    continue;
                // 결과가 True인 경우 카운팅 1 증가
                if (dfs(i, j, graph[i][j]))
                    cnt++;
            }
        }

        System.out.println(cnt);
    }
}
